enzy_htp/structure/structure_constraint/create_constraint.py

```diff
@@ -0,0 +1,310 @@
+            either through Atom() objects or keyword str of Structure().get()
+            combining with a Structure() (e.g.: A.100.CA)
+        target_value:
+            the target dihedral of the constraint
+        topology:
+            the reference topology if keyword str is used for
+            atom_1, atom_2, atom_3, atom_4."""
+    atom_1 = _dispatch_get_key(atom_1, topology)
+    atom_2 = _dispatch_get_key(atom_2, topology)
+    atom_3 = _dispatch_get_key(atom_3, topology)
+    atom_4 = _dispatch_get_key(atom_4, topology)
+    result = DihedralConstraint(
+        atoms=[atom_1, atom_2, atom_3, atom_4],
+        target_value=target_value
+    )
+    result.check_consistent_topology()
+    if params is not None:
+        result.update_params(params)
+
+    return result
+
+def _dispatch_get_key(target: Union[str, Atom, Residue], stru: Structure) -> Union[Atom, Residue]:
+    """apply the key with .get() that support a dispatch of using the
+    object (e.g.: Atom()) itself"""
+    if isinstance(target, str):
+        if not isinstance(stru, Structure):
+            _LOGGER.error("key str types of specification is used but reference "
+                          "topology (i.e.: structure) is not supplied. "
+                          f"(key: {target}, topology: {stru})")
+            raise TypeError
+        target = stru.get(target)
+    return target
+
+# endregion
+
+# region == tools that operations on list/dict of StructureConstraint ==
+def merge_cartesian_freeze(cart_freeze_list: List[CartesianFreeze]) -> CartesianFreeze:
+    """try to merge a list of cartesian freeze into 1 single
+    Cartesian Freeze object. Raise error when they can't merge"""
+
+    # the case only backbone_freeze exists
+    temp_iter = iter(cart_freeze_list)
+    ref_cons = next(temp_iter)
+    try:
+        while ref_cons.constraint_type == "backbone_freeze":
+            ref_cons = next(temp_iter)
+    except StopIteration:
+        _LOGGER.debug(f"Only found specialized backbone_freeze. Returning just BackboneFreeze")
+        return ref_cons
+
+    # check consistency
+    result = CartesianFreeze.clone_from(ref_cons)
+    for cons in cart_freeze_list:
+        if not cons.is_cartesian_freeze:
+            _LOGGER.error(f"this function only take List[CartesianFreeze]. Found: {type(cons)}")
+            raise TypeError
+        if cons.params != ref_cons.params:
+            _LOGGER.error(f"Inconsistent params in the target list. {cons.params} -vs- {ref_cons.params} ")
+            raise ValueError
+        if cons.topology is not ref_cons.topology:
+            _LOGGER.error(f"Inconsistent topology in the target list. {cons.topology} -vs- {ref_cons.topology} ")
+            raise ValueError
+    # merge atoms
+        result.atoms.extend(cons.atoms)
+    result.atoms = list(set(result.atoms))
+    return result
+# endregion
+
+
+def freeze_hydrogen_bonds(sr:StructureRegion, params:Dict=None ) -> List[DistanceConstraint]:
```

Comment:
We can rename it to create_hydrogen_bond_freeze to unify the API

Reply:
sure. done.